### ISSUE TITLE
No code output with imported lower-level most.js modules

### ISSUE THREAD

**Issue by martypdx** [rollup/rollup#1450]
Playing around with rollup bundling mostjs trying to create smaller bundle by importing lower-level modules, I came across this situation (which seems like a bug) where the static analysis in rollup outputs no code:

```js
// main.js
import { fromArray } from 'most/src/source/fromArray';
import { map } from 'most/src/combinator/transform';
import { observe } from 'most/src/combinator/observe';

const stream = fromArray([1,2,3])
const subject = map(x => x + '!!!', stream);
const observer = observe(console.log, subject);

// Without this line, bundle.js is empty except for code comments.
// Uncommented, bundle.js has correct code
// console.log(observer);
```

Rollup and config used: 

```js
// rollup 0.43.0
// rollup.config.js
import resolve from 'rollup-plugin-node-resolve'; // ^3.0.0

export default {
    entry: 'src/main.js',
    format: 'iife',
    plugins: [ 
        resolve({ jsnext: true })
    ],
    dest: 'bundle.js'
};

**Comment by kzc** [comment]
Try: https://github.com/rollup/rollup/issues/786#issuecomment-271934420
```
$ bin/rollup -v
rollup version 0.43.0
```
```
$ cat prom.js
var main = function () {
  console.log('test');
}
let x = new Promise(function (resolve) {
  main();
  resolve();
})
//console.log(x); // uncomment this line
```
```
$ bin/rollup prom.js -f es
//console.log(x); // uncomment this line
```
```
$ cat prom2.js
var main = function () {
  console.log('test');
}
let x = new Promise(function (resolve) {
  main();
  resolve();
})
console.log(x);
```
```
$ bin/rollup prom2.js -f es
var main = function () {
  console.log('test');
};
let x = new Promise(function (resolve) {
  main();
  resolve();
});
console.log(x);
```

**Comment by lukastaegert** [comment]
Hi @martypdx could you please check if you still have this issue with v0.49.3?

**Comment by kzc** [comment]
@lukastaegert It can be verified with `npm install most`. It works.

The Promise fix was required to get the test case to work with v0.48.2, but the test case happens to work with v0.49.3 only because it retains imported code even though the side-effect-free imported symbols are unreferenced. 

Try the following in both versions:
```
// main.js
import { fromArray } from './most/src/source/fromArray.js';
import { map } from './most/src/combinator/transform.js';
import { observe } from './most/src/combinator/observe.js';
```
The difference is due to this regression:
```
var IIFE = (function () {
    function IIFE() {}
    IIFE.prototype.go = 1;
    return IIFE;
}());
```
This example code is dropped in v0.48.2, but retained in v0.49.3.

**Comment by lukastaegert** [comment]
@kzc Thanks for looking into this, good work! So it seems this ticket can be closed but we should make sure we have tests in place that prevent us from breaking this again in the future.  I'll make a note to include something in my next bigger PR that is due to arrive early next week.

The reason the last code example is kept is because side-effects in `IIFE.prototype.go = 1` cannot be determined yet, this is to be expected. However, if this is the only reason why the code is working now, I guess I will need to take another look at what else is not handled properly especially considering the above example seems to be actually side-effect free.

My next PR will seriously refactor many core aspects of call handling so maybe we should revisit this topic again once it is merged.

**Comment by kzc** [comment]
> However, if this is the only reason why the code is working now

@lukastaegert I misstated that. It's *one* of the reasons it's working now - not the *only* reason. If `Promise` is reintroduced in v0.49.3 as being side effect free (which I don't recommend), then the [test](https://github.com/rollup/rollup/issues/1450#issue-236675938) at the top of this ticket would still work.

In any case, this ticket can be closed. The `Promise` fix in v0.49.3 is correct and the `.prototype.property` write issue is a different matter. The side-effect-free IIFE test case can be added to Rollup.
